lib/checkclass.cpp

```diff
@@ -413,30 +413,47 @@ void CheckClass::copyconstructors()
 
     for (const Scope * scope : mSymbolDatabase->classAndStructScopes) {
         std::map<int, const Token*> allocatedVars;
+        std::map<int, const Token*> deallocatedVars;
 
         for (const Function &func : scope->functionList) {
-            if (func.type != FunctionType::eConstructor || !func.functionScope)
-                continue;
-            const Token* tok = func.token->linkAt(1);
-            for (const Token* const end = func.functionScope->bodyStart; tok != end; tok = tok->next()) {
-                if (Token::Match(tok, "%var% ( new") ||
-                    (Token::Match(tok, "%var% ( %name% (") && mSettings->library.getAllocFuncInfo(tok->tokAt(2)))) {
-                    const Variable* var = tok->variable();
-                    if (var && var->isPointer() && var->scope() == scope)
-                        allocatedVars[tok->varId()] = tok;
+            if (func.type == FunctionType::eConstructor && func.functionScope) {
+                // Allocations in constructors
+                const Token* tok = func.token->linkAt(1);
+                for (const Token* const end = func.functionScope->bodyStart; tok != end; tok = tok->next()) {
+                    if (Token::Match(tok, "%var% ( new") ||
+                        (Token::Match(tok, "%var% ( %name% (") && mSettings->library.getAllocFuncInfo(tok->tokAt(2)))) {
+                        const Variable* var = tok->variable();
+                        if (var && var->isPointer() && var->scope() == scope)
+                            allocatedVars[tok->varId()] = tok;
+                    }
                 }
-            }
-            for (const Token* const end = func.functionScope->bodyEnd; tok != end; tok = tok->next()) {
-                if (Token::Match(tok, "%var% = new") ||
-                    (Token::Match(tok, "%var% = %name% (") && mSettings->library.getAllocFuncInfo(tok->tokAt(2)))) {
-                    const Variable* var = tok->variable();
-                    if (var && var->isPointer() && var->scope() == scope && !var->isStatic())
-                        allocatedVars[tok->varId()] = tok;
+                for (const Token* const end = func.functionScope->bodyEnd; tok != end; tok = tok->next()) {
+                    if (Token::Match(tok, "%var% = new") ||
+                        (Token::Match(tok, "%var% = %name% (") && mSettings->library.getAllocFuncInfo(tok->tokAt(2)))) {
+                        const Variable* var = tok->variable();
+                        if (var && var->isPointer() && var->scope() == scope && !var->isStatic())
+                            allocatedVars[tok->varId()] = tok;
+                    }
+                }
+            } else if (func.type == FunctionType::eDestructor && func.functionScope) {
+                // Deallocations in destructors
+                const Token* tok = func.token->linkAt(1);
```

Comment:
```suggestion
                const Token* tok = func.functionScope->bodyStart;
```
For the constructor when we want to loop through the initializer list it makes sense to loop from `func.token->linkAt(1)` but for destructors we just want to loop through the body..